File: spec/cypress/integration/new_messenger/task_bot_spec.js
import helpers from './helpers.js'

function addPackage (app_package) {
  cy.appEval(`
    app = App.first
    app_package = AppPackage.find_by(name: "${app_package}")
    integration = app.app_package_integrations.new()
    integration.app_package = app_package
    integration.save
  `)
}

describe('Task bot Spec', function () {
  beforeEach(() => {
    cy.appEval('ActiveJob::Base.queue_adapter = :test')
    cy.appEval('ActiveJob::Base.queue_adapter.perform_enqueued_at_jobs = true')
    cy.appEval('Redis.current.del("app_user:1:trigger_locked")')
    cy.appEval(`
      require 'app_packages_catalog'
      AppPackagesCatalog.update_all 
    `)
  })

  it('sessionless bot task wait for reply', function () {
    cy.app('clean')
    cy.appScenario('basic').then((basic) => {
      cy.log('basic', JSON.stringify(basic))

      cy.appEval('App.last').then((results) => {
        const appKey = results.key
        cy.log('APP.last', JSON.stringify(results))

        cy.app('bot_task_command', {
          app_key: appKey
        }).then((res) => {
          cy.log('bot task command', JSON.stringify(res))

          cy.wait(2000)
          cy.visit(`/tester/${appKey}?sessionless=true`).then(() => {
            cy.wait(2000)

            cy.get('#messenger-frame iframe')
              .then(function ($iframe) {
                const $body = $iframe.contents().find('body')

                cy.wrap($body).contains('one')
                cy.wrap($body).contains('two')
                cy.wrap($body).contains('tree').then(() => {
                  cy.wait(1000)

                  cy.wrap($body)
                    .xpath('/html/body/div/div/div/div[2]/div[1]/main/div/div/div[4]/div/div/textarea')
                    .type('oeoe \n').then(() => {
                      cy.wrap($body).contains('four')
                    })
                })
              })
          })
        })
      })
    })
  })

  it('sessionless never ask email', function () {
    cy.appScenario('app_bot_settings', { email_requirement: 'never' }).then(() => {
      cy.appEval('App.last').then((results) => {
        const appKey = results.key

        cy.visit(`/tester/${appKey}?sessionless=true&lang=en`).then(() => {

          cy.wait(2000)

          cy.get('#chaskiq-prime').click().then(()=>{
            cy.get('iframe:first')
            .then(function ($iframe) {
              const $body = $iframe.contents().find('body')
              expect($body.html()).to.contain('Start a conversation')

              cy.wrap($body)
                .xpath('/html/body/div/div/div/div[2]/div[1]/div[1]/div/div/div/div[2]/div[2]/a[1]')
                .click()
                .then(() => {

                  cy.wait(2000)
                  cy.wrap($body).find('textarea')
                  .should('have.length', 1)
                  .type('oeoe \n')
      
                  cy.wrap($body).contains('will reply as soon as they can.')
              })
            })
          })
        })
      })
    })
  })

  it('sessionless 2 always ask email , email validation', function () {

    cy.appScenario('app_bot_settings', { email_requirement: 'Always' }).then((results) => {
      addPackage('Qualifier')

      cy.appEval('App.last').then((results) => {
        const appKey = results.key
        cy.visit(`/tester/${appKey}?sessionless=true&lang=en`).then(() => {
          
          cy.wait(2000)
          cy.get('#chaskiq-prime').click()

          cy.wait(2000)
          cy.get('iframe:first')
            .then(function ($iframe) {
              const $body = $iframe.contents().find('body')
              expect($body.html()).to.contain('Start a conversation')

              cy.wrap($body)
                .xpath('/html/body/div/div/div/div[2]/div[1]/div[1]/div/div/div/div[2]/div[2]/a[1]')
                .click()
                .then(() => {
                  cy.wrap($body).then(() => {
                    cy.wait(2000)
                    cy.wrap($body).find('textarea')
                    .should('have.length', 1)
                    .type('oeoe \n').then(() => {
                        cy.wait(2000)

                        cy.wrap($body).contains('will reply as soon as they can.')

                        cy.wrap($body).contains('Are you an existing')

                        cy.wrap($body).contains("Yes, I'm a customer").click().then(() => {
                          cy.wrap($body).contains('by email')
                          cy.wrap($body)
                            .xpath('/html/body/div/div/div/div[2]/div[1]/main/div/div/div[2]/div[1]/div/div/div/div/div/div/form/fieldset/div/div/div/input')
                            .type('John')

                          cy.wrap($body)
                            .xpath('/html/body/div/div/div/div[2]/div[1]/main/div/div/div[2]/div[1]/div/div/div/div/div/div/form/fieldset/div/div/div/div/button')
                            .click()

                          cy.wrap($body).contains('is invalid')

                          cy.wrap($body)
                            .xpath('/html/body/div/div/div/div[2]/div[1]/main/div/div/div[2]/div[1]/div/div/div/div/div/div/form/fieldset/div/div/div/input')
                            .type('John@apple.test')

                          cy.wrap($body)
                            .xpath('/html/body/div/div/div/div[2]/div[1]/main/div/div/div[2]/div[1]/div/div/div/div/div/div/form/fieldset/div/div/div/div/button')
                            .click()

                          cy.wrap($body).contains('Thank you')
                        })
                      })
                })
              })
          })
        })
      })
    })
  })

  it('sessionless 3 never ask email', function () {
    cy.appScenario('app_bot_settings', { email_requirement: 'never' }).then(() => {
      cy.appEval('App.last').then((results) => {
        const appKey = results.key
        cy.visit(`/tester/${appKey}?sessionless=true&lang=en`).then(() => {
         
          cy.wait(2000)

          cy.get('#chaskiq-prime').click()
         
          cy.wait(2000)
          cy.get('iframe:first')
            .then(function ($iframe) {
              const $body = $iframe.contents().find('body')
              expect($body.html()).to.contain('Start a conversation')
              cy.wrap($body)
                .xpath('/html/body/div/div/div/div[2]/div[1]/div[1]/div/div/div/div[2]/div[2]/a[1]')
                .click()
                .then(() => {
                  cy.wrap($body)
                    .xpath('/html/body/div/div/div/div[2]/div[1]/main/div/div/div[4]/div/div/textarea')
                    .should('be.enabled').then(() => {
                      cy.wrap($body)
                        .xpath('/html/body/div/div/div/div[2]/div[1]/main/div/div/div[4]/div/div/textarea')
                        .type('oeoe \n').then(() => {
                          cy.wait(2000)

                          cy.wrap($body).contains('will reply as soon as they can.')

                          cy.wrap($body).contains('oeoe')
                          cy.wrap($body).contains('Are you an existing')
                          cy.wrap($body).contains("I'm a customer").click()
                          cy.wrap($body).contains("reply button: Yes, I'm a customer")
                          //cy.wrap($body).contains("You replied Yes, I'm a customer")
                          cy.wrap($body).contains('you will get a reply from one of our agents')
                        })
                    })
                })
            })
        })
      })
    })
  })
})

describe('start conversation welcome bot', function () {
  it('start_conversation', function () {
    cy.appScenario('basic')

    cy.appEval('App.last').then((results) => {
      const appKey = results.key
      cy.app('new_conversation_bot_task_command', {
        app_key: appKey
      })
    })

    //helpers.openMessenger("?jwt=true", ($body, appKey) => {
    helpers.openMessenger("", ($body, appKey) => {

      cy.wait(6000)
      expect($body.html()).to.contain('Start a conversation')

      cy.wrap($body)
        .xpath('/html/body/div/div/div/div[2]/div[1]/div[1]/div/div/div/div[2]/div[2]/a[1]')
        .click()
        .then(() => {
          cy.wrap($body).contains('see more?').click().then(()=>{
            cy.wait(2000)
            cy.wrap($body).contains('sauper!')

            cy.wrap($body)
              .xpath('/html/body/div/div/div/div[2]/div[1]/main/div/div/div[4]/div/div/textarea')
              .should('be.enabled').then(() => {
                cy.wait(2000)
                cy.wrap($body)
                  .xpath('/html/body/div/div/div/div[2]/div[1]/main/div/div/div[4]/div/div/textarea')
                  .type('oeoe \n').then(() => {
                    cy.wrap($body).contains('oeoe')
                    cy.wrap($body).contains('go to!').click()
                    cy.wrap($body).contains('ah ah !')
                  })
              })
          })

        })
    })

    helpers.openMessenger("", ($body, appKey) => {
      //expect($body.html()).to.contain('Start a conversation')

      console.log($body.html())
      cy.wait(2000)
      cy.wrap($body)
        .xpath('/html/body/div/div/div/div[2]/div[1]/div[1]/div/div/div/div[2]/div[2]/a[1]')
        .click()
        .then(() => {
          cy.wrap($body).contains('see more?').click().then(()=>{
            cy.wait(1500)
            cy.wrap($body).contains('sauper!')
  
            cy.wrap($body)
              .xpath('/html/body/div/div/div/div[2]/div[1]/main/div/div/div[4]/div/div/textarea')
              .should('be.enabled').then(() => {
                cy.wait(2000)
                cy.wrap($body)
                  .xpath('/html/body/div/div/div/div[2]/div[1]/main/div/div/div[4]/div/div/textarea')
                  .type('oeoe \n').then(() => {
                    cy.wrap($body).contains('oeoe')
                    cy.wrap($body).contains('go to!').click()
                    cy.wrap($body).contains('ah ah !')
                  })
              })

          })
        })
    })
  })
})

File: spec/system/messenger/task_bot_spec.rb
require "rails_helper"

RSpec.describe "Task Bot Spec", type: :system, js: true do
  let!(:app) do
    Plugin.restore_plugins_from_fs

    FactoryBot.create(:app, encryption_key: "unodostrescuatro",
                            active_messenger: "true",
                            state: "enabled",
                            domain_url: "*")
  end

  let!(:agent_role) do
    app.add_admin({ email: "miguel.michelson@cience.com", name: "test agent", role: "admin" })
  end

  before do
    ActiveJob::Base.queue_adapter = :test
    ActiveJob::Base.queue_adapter.perform_enqueued_at_jobs = true
    # Additional setup, including Redis and AppPackagesCatalog updates
    # ...
  end

  before :each do
    ActiveJob::Base.queue_adapter = :test
    ActiveJob::Base.queue_adapter.perform_enqueued_at_jobs = true
    Redis.current.del("app_user:1:trigger_locked")

    require "app_packages_catalog"
    AppPackagesCatalog.update_all
  end

  it "sessionless bot task wait for reply" do
    app.update({ inline_new_conversations: false })
    bot_task_command({
                       app_key: app.key
                     })

    open_new_messenger({ disable_open: true, params: { lang: "en", old_embed: nil } }, sessionless: true) do |page|
      expect(page).to have_content("one")
      expect(page).to have_content("two")
      expect(page).to have_content("tree")
      #  # Typing into textarea and checking response

      page.find("textarea").set("oeoe").native.send_keys(:return)

      expect(page).to have_content("oeoe")

      expect(page).to have_content("four")
    end
  end

  it "sessionless never ask email'" do
    app_bot_settings({ email_requirement: "never" })

    open_new_messenger({ params: { lang: "en" } }, sessionless: true) do |page|
      click_link "Start a conversation"
      #  .xpath('/html/body/div/div/div/div[2]/div[1]/div[1]/div/div/div/div[2]/div[2]/a[1]').click()
      # Typing into textarea and checking response
      page.find("textarea").set("oeoe").native.send_keys(:return)
      expect(page).to have_content("will reply as soon as they can.")
    end
  end

  it "sessionless 2 always ask email , email validation" do
    app_bot_settings({ email_requirement: "Always" })

    add_app_package("Qualifier")

    open_new_messenger({ params: { lang: "en" } }, sessionless: true) do |page|
      click_link "Start a conversation"
      #  .xpath('/html/body/div/div/div/div[2]/div[1]/div[1]/div/div/div/div[2]/div[2]/a[1]').click()
      # Typing into textarea and checking response
      page.find("textarea").set("oeoe").native.send_keys(:return)
      expect(page).to have_content("will reply as soon as they can.")

      expect(page).to have_content("Are you an existing")
      expect(page).to have_content("Yes, I'm a customer")
      click_button("Yes, I'm a customer")
      # click_button("Are you an existing")
      expect(page).to have_content("by email")
    end
  end

  it "sessionless always ask email, email validation" do
    # Setup app with email requirement always
    app_bot_settings({ email_requirement: "Always" })

    # Add the 'Qualifier' package or similar setup
    add_app_package("Qualifier")

    open_new_messenger({ params: { lang: "en" } }, sessionless: true) do |page|
      expect(page).to have_content("Start a conversation")
      click_link "Start a conversation"

      textarea = find("textarea")
      textarea.set("oeoe").native.send_keys(:return)

      expect(page).to have_content("will reply as soon as they can.")
      expect(page).to have_content("Are you an existing")

      page.find("Yes, I'm a customer").click # Update this selector based on actual content
      expect(page).to have_content("by email")

      input_field = page.find(:xpath, "/html/body/div/div/div/div[2]/div[1]/main/div/div/div[2]/div[1]/div/div/div/div/div/div/form/fieldset/div/div/div/input")
      input_field.set("John")
      page.find(:xpath, "/html/body/div/div/div/div[2]/div[1]/main/div/div/div[2]/div[1]/div/div/div/div/div/div/form/fieldset/div/div/div/div/button").click

      expect(page).to have_content("is invalid")

      input_field.set("John@apple.test")
      page.find(:xpath, "/html/body/div/div/div/div[2]/div[1]/main/div/div/div[2]/div[1]/div/div/div/div/div/div/form/fieldset/div/div/div/div/button").click

      expect(page).to have_content("Thank you")
    end
  end

  it "sessionless never asks for email" do
    # Setup app with email requirement set to 'never'
    app_bot_settings({ email_requirement: "never" })

    open_new_messenger({ params: { lang: "en" } }, sessionless: true) do |page|
      expect(page).to have_content("Start a conversation")
      click_link("Start a conversation")

      textarea = page.find("textarea")
      expect(textarea).to be_visible
      textarea.set("oeoe").native.send_keys(:return)
      sleep(2) # Wait for response

      expect(page).to have_content("will reply as soon as they can.")
      expect(page).to have_content("oeoe")
      expect(page).to have_content("Are you an existing")
      expect(page).to have_content("I'm a customer") # Update the text based on actual content
      # Additional assertions or interactions if needed
      page.click_button("I'm a customer")
      expect(page).to have_content("reply button: Yes, I'm a customer")
      expect(page).to have_content("you will get a reply from one of our agents")
    end
  end

  describe "Start Conversation Welcome Bot", type: :system, js: true do
    it "start_conversation" do
      new_conversation_bot_task_command({
                                          app_key: app.key
                                        })

      open_new_messenger({ params: { lang: "en" } }, sessionless: false) do |page|
        expect(page).to have_content("Start a conversation")
        page.find("see more?").click
        expect(page).to have_content("see more?")
        page.find("see more?").click
        sleep(2)
        expect(page).to have_content("sauper!")

        page.find("textarea").set("oeoe").native.send_keys(:return)
        expect(page).to have_content("oeoe")
        page.find("go to!").click
        expect(page).to have_content("ah ah !")
      end
    end
  end
end
